### DISCUSSION TITLE
API proposal

### DISCUSSION THREAD

**Discussion by IlianPihlajamaa** [IlianPihlajamaa/ModeCouplingTheory.jl#6] [General]
Here is my current idea for an API:

For the standard equations:
```julia
kernel = MyKernel(...)
problem = MCTProblem(parameters.., initial_conditions..., kernel) 
solver = MySolver(solver_parameters...) 
t, F, K = solve(problem, solver)
```

For the tagged-particle equations this is basically the same:
```julia
kernel = TaggedParticleMCTKernel (...)
problem = TaggedParticleMCTProblem(parameters.., initial_conditions..., kernel) 
solver = MySolver(solver_parameters...) 
t, F, K = solve(problem, solver)
```
meaning that the `TaggedParticleMCTProblem` function handles "stacking" the vectors `F` and `F_s` (collective and self intermediate scattering function) and returns perhaps a `LinearMCTProblem`. That way, the solver will never notice it isn't just solving a large standard `LinearMCTProblem`. We can easily make `F` and `K` a 2-tuple containing `(F, Fs)`, in order to make it easy for users.

**Comment by thvoigtmann** [comment]
I see two possible approaches: (1) "stacking" of models into bigger models that are solved simultaneously; (2) defining models that take as one of their parameters a given numerical solution of another model, together with some metadata in order to make sense of it.

Pro (1):
- allows the solver to throw away intermediates; e.g. one can do in-place decimization of the time grid
- easier to synchronize solver parameters to match

Pro (2):
- easier to implement the later extension of previously stored solutions
- potentially easier to parallelize (e.g. multiple tagged particles attached to the same base model)

Having followed (1) myself for the past ten or so years, I'm beginning to lean towards (2)...

So I'd (similar to the example documentation right now)
```julia
kernel = MyKernel(...)
problem = MCTProblem(parameters..., initial_conditions..., kernel)
solver = MySolver(solver_parameters...)
fqt = solve(problem, solver)
```
Where `fqt` is a struct storing the data (see below). And then
```julia
kernel_s = TaggedParticleMCTKernel(parameters..., fqt)
problem_s = TaggedParticleMCTProblem(parameters..., initial_conditions..., kernel_s)
solver_s = MySolver(solver_parameters...)
fsqt = solve(problem_s, solver_s)
```

The `TaggedParticleMCTKernel` will need to know the q-grid of `fqt` (and the like), so let's say `fqt` contains a data field that stores that, then `TaggedParticleMCTKernel` can read the information from there (or implementations could allow interpolation from the given q-grid of the `fqt` solution to whatever they use). They can throw if there is no match.

`solve` will need to know the solver information that has been used to obtain `fqt`. (Mostly the time grid for now.) It would take them from there and ignore any conflicting parameters given upon initialization (presumably raising a warning). For example, in the `FuchsSolver` we want the size of the grid blocks and the time steps to match; but accuracy etc. could differ. Implementations could even choose to provide interpolation.

It's not exactly clear to me how to best fit this into Julia, but let's say
```julia
abstract type BlockedTimeGridSolver end
struct MySolver <: BlockedTimeGridSolver
    """ solver implementation here """
end
abstract type MCTSolutions end
mutable struct BlockedTimeGridSolutions <: MCTSolutions
    solver::BlockedTimeGridSolver
    model::MCTModel
    """ t, F, K of appropriate type here """
end
```
And say that we have the signature `solve(::MCTProblem, ::MySolver)::BlockedTimeGridSolutions`. For the tagged-particle kernel I would then implement a function `update_kernel!(::MemoryKernel, ::MCTSolver, t_index)` to be called by the solver every time it starts a new time step. That function can update the tagged-particle vertices (let's interpret them as time-dependent vertices since they are static vertices multiplied with a collective correlation function), so that the `evaluate_kernel` functions can efficiently iterate the solution. The `update_kernel!` function would check whether the given solver parameters are compatible with the one the `kernel_s` stores from `fqt` and do whatever seems reasonable (continue, throw, interpolate).

The `update_kernel!` probably will need a wrapper function `f(indices..., fqt)` of some sort to pick out the right elements of the collective correlator.

It feels a bit dirty to make all these requirements run-time errors/warnings, instead of catching them with proper abstract types and subtypes, I admit. And I admit it gets a bit messy in my head without coding right now...

**Comment by IlianPihlajamaa** [comment]
There is a significant performance penalty for option (2) though (at least for schematic models)... I find that solving for a stacked schematic F2 model is roughly twice as fast as solving two scalar schematic models (still using tDict, which accounted for 4% of the runtime). I checked that the results are the same. 

code to reproduce

```julia
using ModeCouplingTheory, BenchmarkTools, StaticArrays, LinearAlgebra

import ModeCouplingTheory.MemoryKernel
struct TaggedSchematicKernel <: MemoryKernel
    λ::Float64
    F::Vector{Float64}
    tDict::Dict{Float64, Int64}
end

import ModeCouplingTheory.evaluate_kernel
@inline function evaluate_kernel(kernel::TaggedSchematicKernel, Fs, t)
    @inbounds it = kernel.tDict[t] # find the correct index corresponding to t
    return @inbounds kernel.F[it]*Fs*kernel.λ
end

## Option 2
ts, Fs, Ks = @btime begin 
    F0 = 1.0; ∂F0 = 0.0; α = 0.0; β = 1.0; γ = 1.0
    kernel = SchematicF2Kernel(3.98)
    problem = LinearMCTProblem(α, β, γ, F0, ∂F0, kernel)
    solver = FuchsSolver(problem) 
    t, F, K = solve(problem, solver, kernel); # 836.700 μs (39 allocations: 366.58 KiB)

    tDict = Dict(zip(t, eachindex(t)))
    taggedkernel = TaggedSchematicKernel(3.9, F, tDict)
    taggedproblem = LinearMCTProblem(α, β, γ, F0, ∂F0, taggedkernel)
    taggedsolver = FuchsSolver(taggedproblem)
    ts, Fs, Ks = solve(taggedproblem, taggedsolver, taggedkernel) # 952.300 μs (39 allocations: 366.58 KiB)
end


struct TaggedStackedKernel <: MemoryKernel
    λ::SVector{2, Float64}
end

@inline function evaluate_kernel(kernel::TaggedStackedKernel, F, t)
    return Diagonal(SVector{2, Float64}(kernel.λ[1]*F[1]^2, kernel.λ[2]*F[1]*F[2]))
end

## Option 1
t, F, K = @btime begin 
    F0 = @SVector(ones(2)); ∂F0 = @SVector(zeros(2)); α = 0.0; β =  @SVector(ones(2)); γ = @SVector(ones(2))
    kernel = TaggedStackedKernel(@SVector([3.98, 3.9]))
    problem = LinearMCTProblem(α, β, γ, F0, ∂F0, kernel)
    solver = FuchsSolver(problem)
    t, F, K = solve(problem, solver, kernel) #935.800 μs (43 allocations: 816.75 KiB)
end;
```
prints

```
  1.922 ms (1173 allocations: 902.39 KiB)
  963.300 μs (603 allocations: 848.70 KiB)
```

**Comment by thvoigtmann** [comment]
I don't think that the comparison is entirely fair: you are hard-coding the F12+Sjögren case as Option 1, but for a generic code we would need to find a way of stacking models that the user defined without literally copying the kernel code. This will always have some penalty for those simple models, but I expect the overhead to be small enough for full-blown MCT models.

Just for fun, I tried
```julia
struct TaggedStackedKernel <: MemoryKernel
    λs::Float64
    base_kernel
end

@inline function evaluate_kernel(kernel::TaggedStackedKernel, F, t)
    return Diagonal(SVector{2, Float64}(evaluate_kernel(kernel.base_kernel, F[1], t), kernel.λs*F[1]*F[2]))
end
```
which changes the API to
```julia
kernel = TaggedStackedKernel(3.9,SchematicF2Kernel(3.98))
```

And this turns out to be horrible (nearly 10 times slower).

Also, just re-evaluating the base model's kernel every time we solve another tagged-particle is perhaps not the way to go :)

**Comment by thvoigtmann** [comment]
So, afterthought: providing specific code for F12+Sjögren as you just did as a special more performant case, would be my suggestion ;) It'll mess up some functions involved in calculating the bifurcation points because as a two-component model, F12+Sjögren does not have an irreducible stability matrix, but that is not going to be a major application of the code.

BTW, I have a request: can we change the name of the schematic-model parameters from λ to v?

**Reply by IlianPihlajamaa** [reply]
> BTW, I have a request: can we change the name of the schematic-model parameters from λ to v?

done. We should perhaps think about the naming of the schematic memory kernels and which ones we want to make visible to users. I implemented the ones that are there now for testing purposes.

**Comment by IlianPihlajamaa** [comment]
I think what you find is mainly due to the fact that your `base_kernel` field is untyped, so julia has to box it. Changing to the code below restores my original timings (roughly)

```julia
using ModeCouplingTheory, BenchmarkTools, StaticArrays, LinearAlgebra

import ModeCouplingTheory.MemoryKernel
struct TaggedSchematicKernel <: MemoryKernel
    λ::Float64
    F::Vector{Float64}
    tDict::Dict{Float64, Int64}
end

import ModeCouplingTheory.evaluate_kernel
@inline function evaluate_kernel(kernel::TaggedSchematicKernel, Fs, t)
    @inbounds it = kernel.tDict[t] # find the correct index corresponding to t
    return @inbounds kernel.F[it]*Fs*kernel.λ
end

ts, Fs, Ks =  begin 
    F0 = 1.0; ∂F0 = 0.0; α = 0.0; β = 1.0; γ = 1.0
    kernel = SchematicF2Kernel(3.98)
    problem = LinearMCTProblem(α, β, γ, F0, ∂F0, kernel)
    solver = FuchsSolver(problem)
    t, F, K = @btime solve(problem, solver, kernel);

    tDict = Dict(zip(t, eachindex(t)))
    taggedkernel = TaggedSchematicKernel(3.9, F, tDict)
    taggedproblem = LinearMCTProblem(α, β, γ, F0, ∂F0, taggedkernel)
    taggedsolver = FuchsSolver(taggedproblem)
    ts, Fs, Ks =  @btime solve(taggedproblem, taggedsolver, taggedkernel)
end


struct TaggedStackedKernel{S} <: MemoryKernel
    λs::Float64
    base_kernel::S
end

@inline function evaluate_kernel(kernel::TaggedStackedKernel, F, t)
    return Diagonal(SVector{2, Float64}(evaluate_kernel(kernel.base_kernel, F[1], t), kernel.λs*F[1]*F[2]))
end

t, F, K = begin 
    F0 = @SVector(ones(2)); ∂F0 = @SVector(zeros(2)); α = 0.0; β =  @SVector(ones(2)); γ = @SVector(ones(2))
    kernel = SchematicF2Kernel(3.98)
    kernel_s = TaggedStackedKernel(3.9, kernel)
    problem = LinearMCTProblem(α, β, γ, F0, ∂F0, kernel_s)
    solver = FuchsSolver(problem)
    t, F, K = @btime solve(problem, solver, kernel_s)
end;

```

**Comment by thvoigtmann** [comment]
Yes, I should have spotted that. I don't have time right now, but it would be interesting to see timings for the more realistic case of a HS model with a tagged particle.

**Comment by IlianPihlajamaa** [comment]
I did some tests with a larger model (50-k points and a `O(Nk^3)` memory kernel) and the timings between both methods are very close (sequentially solving them I find is marginally faster, perhaps due to cache efficiency, but it is less then 5-10% and variable). Stacking the kernels is a lot easier though. The implementation is roughly this

```julia
struct CollectiveKernel{V} <: MemoryKernel
    vertices²::V
end

struct TaggedKernel{V} <: MemoryKernel
    vertices²::V
end

struct StackedKernel{C,S} <: MemoryKernel
    kernel_c::C
    kernel_s::S
end

function evaluate_kernel(out, kernel::StackedKernel, F, t)
    Nk = length(F) ÷ 2 # this should be a field of the kernel probably (if more than two models are stacked)
    @views evaluate_kernel(kernel.kernel_c, F[1:Nk], t) # the collective F is given by the first Nk elements of F
    evaluate_kernel!(kernel.kernel_s, F, t) # the tagged kernel needs both the collective and tagged F.
end
```
In this way, arbitrary kernels can be stacked without any other boilerplate code. We can create a `combine_problems` function that automatically stacks the parameters and initial conditions of two `LinearMCTProblem`s and returns a bigger `LinearMCTProblem` (or -`Equation`) with the correct memory kernel and parameters. I think this is a very clean solution. It is also a natural way to solve GMCT-like equations if it ever comes to that (they cannot be solved sequentially and have to be stacked). 

I am unsure how often it is necessary in practice, but a workaround for finding the tagged dynamics for already solved `Problems` is the following:

We can define an `InterpolatingMemoryKernel` that takes a memory kernel (measured from simulations or taken from an earlier numerical solution), and linearly interpolates in time to find $K(t)$. For an earlier solved Problem, one can then use the `InterpolatingMemoryKernel` as the collective kernel and a `TaggedMemoryKernel` for the tagged dynamics, stacking them. Since interpolation for all k at some t is a `O(Nk)` operation, it should not have any overhead when compared to a `O(Nk^2)` memory kernel evaluation. For schematic models where this overhead does exist I don't think anybody would really bother in the first place since we solve such equations in milliseconds anyway.

Your suggestion for the sequential solution is also possible of course, but I still don't really see when it would be of great advantage. If someone solved some MCTequation a long time ago for the collective dynamics, and saved the result to disk, probably the MCTsolution object itself is lost, and either needs to be reconstructed in some way (maybe guessing the solver parameters, which may be dangerous). In that case it will probably be safer to just accept the factor of 2 performance loss and re-solve the collective equation. For multiple tagged particles, we can stack an arbitrary number of memory kernels and evaluate them with the pattern above.

**Reply by thvoigtmann** [reply]
> I think this is a very clean solution. It is also a natural way to solve GMCT-like equations if it ever comes to that (they cannot be solved sequentially and have to be stacked).

OK, that convinced me :)

Semantic comment: I would stack models, not kernels. Rationale: we will want to stack things together that need to call different solvers. For example, a mixture problem with a scalar tagged particle problem. They need differently typed solvers. Also the MSD I would implement as a different solver to avoid performance penalties. (Or at least in a way that the needed if's can be optimized away by LLVM.)

So we could go for a `MCTModelStack <: MCTModel` that essentially takes a list of `MCTModel`s and stores some data fields on how to construct the views into the big correlator array that each individual model needs. (Views, a clean solution, I like it.) And the `evaluate_kernel` dispatched on `KernelStack{MCTModelStack}` just calls the `evaluate_kernel` functions dispatched on the individual models in order.

**Comment by IlianPihlajamaa** [comment]
After thinking on it more, I am sort of reverting on my belief that the stacking procedure is the way to go. I have not thought of a way to implement it without any major performance penalties or requiring the users to jump through weird loops in order to "link" the equations together. I think something along the lines of your earlier suggestion:

> So I'd (similar to the example documentation right now)
> 
> kernel = MyKernel(...)
> problem = MCTProblem(parameters..., initial_conditions..., kernel)
> solver = MySolver(solver_parameters...)
> fqt = solve(problem, solver)
> Where fqt is a struct storing the data (see below). And then
> 
> kernel_s = TaggedParticleMCTKernel(parameters..., fqt)
> problem_s = TaggedParticleMCTProblem(parameters..., initial_conditions..., kernel_s)
> solver_s = MySolver(solver_parameters...)
> fsqt = solve(problem_s, solver_s)

Is by far the easiest to correctly implement (preventing bugs), and the most insightful for users, such as perhaps an experimentalist who just wants a quick MSD. As you mentioned, it also natively allows different convergence criteria for the different equations, which is really hard to do in the stacked case. The problems presented earlier:

- Performance of schematic tagged models: I think that users who really cares about the performance of small schematic models are usually savvy enough to hand-code the full combined kernel (which performance-wise we cannot beat in a generic way, so they will do so anyway). For the super common ones we can provide preconstructed kernels. The performance of big models actually seems faster when the equations are solved sequentially, which to me is more important
- All other models (except GMCT) that I can think of have a linear dependency chain, without loops, so they can all be solved in this way. This holds for tagged dynamics, MSD, susceptibilities, and "exponential GMCT". Am I missing any?
- For the specific case of GMCT, once the individual kernel evaluations have been implemented, it is really easy to stack them manually. This also allows for a more fine grained control of the order in which they are evaluated and the way they are stored in memory, which is very important for these huge models.
